From 86845a85bf79e7d83f4c5fbf5362eefebb8f07e3 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 12 May 2023 17:34:38 +0200
Subject: [PATCH 0396/1695] drivers: stm32_rng: ensure conditional reset
 sequence is done

Add a check on RNG_CR_CONDRST being cleared before continuing the
program to ensure that the conditional reset sequence is done.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 66a6b14ee5505fda798470eaaa12510c3b444bbd drivers: stm32_rng: ensure conditional reset sequence is done]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I08db57db5f77e1f7101e9c7accc8f0814dfc9a07
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/306295
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_rng.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index 43133b024..1f0b43f8a 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -320,10 +320,16 @@ static TEE_Result init_rng(void)
 		if(stm32_rng->health_test_conf)
 			io_write32(rng_base + RNG_HTCR, stm32_rng->health_test_conf);
 
-		/* No need to wait for RNG_CR_CONDRST toggle as we enable clk */
 		io_clrsetbits32(rng_base + RNG_CR, RNG_CR_CONDRST,
 				RNG_CR_RNGEN);
 
+		timeout_ref = timeout_init_us(RNG_READY_TIMEOUT_US);
+		while ((io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST))
+			if (timeout_elapsed(timeout_ref))
+				break;
+		if ((io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST))
+			panic();
+
 		DMSG("RNG control register %#"PRIx32,
 		     io_read32(rng_base + RNG_CR));
 		DMSG("RNG noise source control register %#"PRIx32,
@@ -439,6 +445,8 @@ static TEE_Result stm32_rng_pm_resume(void)
 	io_write32(base + RNG_SR, 0);
 
 	if (stm32_rng->ddata->has_cond_reset) {
+		uint64_t timeout_ref = 0;
+
 		/*
 		 * Configuration must be set in the same access that sets
 		 * RNG_CR_CONDRST bit. Otherwise, the configuration setting is
@@ -452,6 +460,13 @@ static TEE_Result stm32_rng_pm_resume(void)
 		io_write32(base + RNG_HTCR, stm32_rng->pm_health);
 
 		io_clrsetbits32(base + RNG_CR, RNG_CR_CONDRST, RNG_CR_RNGEN);
+
+		timeout_ref = timeout_init_us(RNG_READY_TIMEOUT_US);
+		while ((io_read32(base + RNG_CR) & RNG_CR_CONDRST))
+			if (timeout_elapsed(timeout_ref))
+				break;
+		if ((io_read32(base + RNG_CR) & RNG_CR_CONDRST))
+			panic();
 	} else {
 		io_write32(base + RNG_CR, RNG_CR_RNGEN | stm32_rng->pm_cr);
 	}
-- 
2.47.2

